model/flow/header.go

```diff
@@ -80,7 +86,50 @@ func (h Header) ID() Identifier {
 	if h.Timestamp.Location() != time.UTC {
 		h.Timestamp = h.Timestamp.UTC()
 	}
-	return MakeID(h)
+
+	mutexHeader.Lock()
+
+	// unlock at the return
+	defer mutexHeader.Unlock()
+
+	// compare these elements individually
+	if prevHeader.ParentVoterIDs != nil &&
+		prevHeader.ParentVoterSigData != nil &&
+		prevHeader.ProposerSigData != nil &&
+		len(h.ParentVoterIDs) == len(prevHeader.ParentVoterIDs) &&
+		len(h.ParentVoterSigData) == len(prevHeader.ParentVoterSigData) &&
+		len(h.ProposerSigData) == len(prevHeader.ProposerSigData) {
+		bNotEqual := false
+
+		for i, v := range h.ParentVoterIDs {
+			if v == prevHeader.ParentVoterIDs[i] {
+				continue
+			}
+			bNotEqual = true
+			break
+		}
+		if !bNotEqual &&
+			h.ChainID == prevHeader.ChainID &&
+			h.Timestamp == prevHeader.Timestamp &&
+			h.Height == prevHeader.Height &&
+			h.ParentID == prevHeader.ParentID &&
+			h.View == prevHeader.View &&
+			h.PayloadHash == prevHeader.PayloadHash &&
+			bytes.Equal(h.ProposerSigData, prevHeader.ProposerSigData) &&
+			bytes.Equal(h.ParentVoterSigData, prevHeader.ParentVoterSigData) &&
+			h.ProposerID == prevHeader.ProposerID {
+
+			// cache hit, return the previous identifier
+			return previdHeader
+		}
+	}
```

Comment:
Could we write a micro-benchmark for this so we keep an eye on the performance of MakeID?

Reply:
yeah I was thinking about the same thing, I think this work assumes the calls to a header hash are repeated sequentially but the overhead of mutex lock and equality checks might also not be negligible. a micro-benchmark code or tests would help to show the correctness and performance of this as a separate componenet.

Reply:
+1

Reply:
I can do this with an external tool.  I'll make an issue for it, and add it to the list of performance benchmarks I'm keeping.